hellsemble/hellsemble.py

```diff
@@ -91,21 +91,59 @@ def fit(
         Returns:
             Hellsemble: The fitted Hellsemble model.
         """
-        if isinstance(X, pd.DataFrame):
-            X = X.values
+        (
+            X_train,
+            X_validation,
+            y_train,
+            y_validation,
+            train_idx,
+            validation_idx,
+        ) = train_test_split(
+            X,
+            y,
+            np.arange(len(X)),
+            test_size=validation_size,
+            random_state=seed,
+        )
+        if isinstance(X_train, pd.DataFrame):
+            X_train = X_train.values
         if self.mode == "greedy":
-            self.__fitting_history = self.__fit_estimators_greedy(
-                X, y, validation_size, stopping_threshold, seed
+            self.__fitting_history, self.__validation_fitting_history = (
+                self.__fit_estimators_greedy(
+                    X_train,
+                    y_train,
+                    X_validation,
+                    y_validation,
+                    stopping_threshold,
+                )
             )
         else:
-            self.estimators, self.__fitting_history = (
-                self.__fit_estimators_sequential(
-                    X, y, validation_size, stopping_threshold, seed
-                )
+            (
+                self.__fitting_history,
+                self.__validation_fitting_history,
+            ) = self.__fit_estimators_sequential(
+                X_train,
+                y_train,
+                X_validation,
+                y_validation,
+                stopping_threshold,
             )
+
         if len(self.estimators) > 1:
+
+            n_estimators = len(self.__fitting_history)
+            n_total = X.shape[0]
+            full_fitting_history = []
+            for i in range(n_estimators):
+                full_mask = np.zeros(n_total, dtype=bool)
```

Comment:
I couldn't tell if using both train and val data for routing model is a good idea or not. Using train may tend to overfit (ensemble models are better on train so labels are more adequate). Maybe we should discuss it